### PR TITLE
Records are not supported in qute template

### LINKED ISSUE #461: Records are not supported in qute template
With a Java 17 project, define a record such as 
```
package org.acme;

public record Item(String name, java.math.BigDecimal price) {}
```
When used in a qute template, you don't get any completion for name or price:
```
{@org.acme.Item item}
{item.|}
```
Screenshot 2021-11-30 at 19 54 20

For some reason, the duplicate methods bug (#460) shows up, even though the record doesn't extend a custom object.

### PR DESCRIPTION
**Pull request by angelozerr**
Records are not supported in qute template

Fixes #461

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by fbricon** [comment]
You'd need  a Java 15 install on the CI for the tests to pass. (Java 17 would be best though)

**Comment by angelozerr** [comment]
> You'd need a Java 15 install on the CI for the tests to pass. (Java 17 would be best though)

Do you know how to do that?

Is it working for you?

**Comment by rgrunber** [comment]
Is it worth creating something like https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/TestVMType.java (that uses `fakejdk/${version}/rtstubs.jar` to simulate various JDKs ?

**Comment by fbricon** [comment]
@rgrunber ideally yes. 
But for now the easiest way to make the build run `record` tests is to use a JDK 17 runtime. I can provide a PR for that.
Unless you prefer to got the rtstubs route first

**Comment by rgrunber** [comment]
Ok, let's just bump the JDK to 16 (or 17), whichever is easiest. @angelozerr , don't forget about the Jenkinsfile (still references 15, and records become final in 16).

**Comment by angelozerr** [comment]
> don't forget about the Jenkinsfile (still references 15, and records become final in 16).

Indeed I noticed that now :) I fixed it. Hope CI build will work.

File: Jenkinsfile [hunk 1]
```diff
@@ -13,7 +13,7 @@ pipeline {
       when { not { equals expected: '', actual: VERSION } }
       steps {
         sh '''
-          export JAVA_HOME="${NATIVE_TOOLS}${SEP}jdk11_last"
+          export JAVA_HOME="${NATIVE_TOOLS}${SEP}jdk16_last"
```
Comment (fbricon) [inline Jenkinsfile:16]: openjdk17_last

File: Jenkinsfile [hunk 2]
```diff
@@ -26,7 +26,7 @@ pipeline {
       steps {
         sh '''
           WORKSPACE_SAV=${WORKSPACE}
-          export JAVA_HOME="${NATIVE_TOOLS}${SEP}jdk11_last"
+          export JAVA_HOME="${NATIVE_TOOLS}${SEP}jdk16_last"
```
Comment (fbricon) [inline Jenkinsfile:29]: openjdk17_last

File: Jenkinsfile [hunk 3]
```diff
@@ -56,7 +56,7 @@ pipeline {
       steps {
         sh '''
           mvnFlags="-B -ntp -DaltSnapshotDeploymentRepository=origin-repository.jboss.org"
-          export JAVA_HOME="${NATIVE_TOOLS}${SEP}jdk11_last"
+          export JAVA_HOME="${NATIVE_TOOLS}${SEP}jdk16_last"
```
Comment (fbricon) [inline Jenkinsfile:59]: openjdk17_last

File: qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-java16/.classpath [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="src" output="target/test-classes" path="src/test/java">
+		<attributes>
+			<attribute name="optional" value="true"/>
+			<attribute name="maven.pomderived" value="true"/>
+			<attribute name="test" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-16">
```
Comment (fbricon) [inline qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-java16/.classpath:21]: JavaSE-17

File: qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-java16/pom.xml [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  <version>1.0.0-SNAPSHOT</version>
+  <properties>
+    <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
+    <quarkus.platform.group-id>io.quarkus</quarkus.platform.group-id>
+    <quarkus.platform.version>2.4.2.Final</quarkus.platform.version>
+    <surefire-plugin.version>3.0.0-M5</surefire-plugin.version>
+    <compiler-plugin.version>3.8.0</compiler-plugin.version>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <maven.compiler.source>16</maven.compiler.source>
+    <maven.compiler.target>16</maven.compiler.target>
```
Comment (fbricon) [inline qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-java16/pom.xml:16]: replace both with `<maven.compiler.release>17</maven.compiler.release>`

File: qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/internal/QuteProjectTest.java [hunk 6]
```diff
@@ -58,6 +58,7 @@
 
 		public static String qute_quickstart = "qute-quickstart";
 
+		public static String qute_java16 = "qute-java16";
```
Comment (fbricon) [inline qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/internal/QuteProjectTest.java:61]: 17

**Review by fbricon — CHANGES_REQUESTED** [review]

**Comment by fbricon** [comment]
you need to update the gradle test project's version to 7.3 (for Java 17 compat)

**Comment by angelozerr** [comment]
> you need to update the gradle test project's version to 7.3 (for Java 17 compat)

Never done that, I need to study how to do that.

**Comment by fbricon** [comment]
Doesn't work at all. Test failures is most likely caused by trying to collect fields via type.getFields() instead of type.getRecordComponents(), in QuteSupportForTemplate.getResolvedJavaType

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java [hunk 7]
```diff
@@ -368,6 +368,9 @@ private static boolean isIterable(IType iterableType, IProgressMonitor monitor)
 	}
```
Comment (fbricon) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java:369]: L283, it's missing something like
```java
if(type.isRecord()) {
	for (IField field : type.getRecordComponents()) {
		//All record components are valid
		JavaFieldInfo info = new JavaFieldInfo();
		info.setSignature(typeResolver.resolveFieldSignature(field));
		fieldsInfo.add(info);
	}
}
```

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java [hunk 8]
```diff
@@ -368,6 +368,9 @@ private static boolean isIterable(IType iterableType, IProgressMonitor monitor)
 	}
 
 	private static boolean isValidField(IField field) throws JavaModelException {
+		if (field.isRecordComponent()) {
+			return true;
+		}
```
Comment (fbricon) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java:373]: not necessary, given the previous comment

File: qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetResolvedJavaTypeTest.java [hunk 9]
```diff
@@ -145,9 +145,30 @@ public void methodArray() throws Exception {
...
+				QuteMavenProjectName.qute_java17);
+		ResolvedJavaTypeInfo result = QuteSupportForTemplate.getInstance().getResolvedJavaType(params, getJDTUtils(),
+				new NullProgressMonitor());
+		Assert.assertEquals("org.acme.qute.RecordItem", result.getSignature());
+		Assert.assertFalse(result.isIterable());
+
+		Assert.assertNotNull(result.getFields());
+		Assert.assertEquals(2, result.getFields().size());
+		Assert.assertEquals("name", result.getFields().get(0).getName());
+		Assert.assertEquals("java.lang.String", result.getFields().get(0).getSourceType());
```
Comment (fbricon) [inline qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetResolvedJavaTypeTest.java:169]: > Assert.assertEquals("java.lang.String", result.getFields().get(0).getType());

works better

File: qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetResolvedJavaTypeTest.java [hunk 10]
```diff
@@ -145,9 +145,30 @@ public void methodArray() throws Exception {
...
+		Assert.assertEquals("price", result.getFields().get(1).getName());
+		Assert.assertEquals("java.math.BigDecimal", result.getFields().get(1).getSourceType());
```
Comment (fbricon) [inline qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetResolvedJavaTypeTest.java:171]: > Assert.assertEquals("java.math.BigDecimal", result.getFields().get(1).getType());

works better

**Review by fbricon — CHANGES_REQUESTED** [review]

File: qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetResolvedJavaTypeTest.java (hunk 9 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetResolvedJavaTypeTest.java]: fixed

File: qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetResolvedJavaTypeTest.java (hunk 10 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetResolvedJavaTypeTest.java]: thanks fixed

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java (hunk 8 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java]: fixed

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java (hunk 7 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java]: thanks so much, fixed

File: qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/internal/QuteProjectTest.java (hunk 6 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/internal/QuteProjectTest.java]: fixed

File: qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-java16/pom.xml (hunk 5 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-java16/pom.xml]: fixed

File: qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-java16/.classpath (hunk 4 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-java16/.classpath]: fixed

File: Jenkinsfile (hunk 3 above)
Comment (angelozerr) [reply Jenkinsfile]: fixed

File: Jenkinsfile (hunk 2 above)
Comment (angelozerr) [reply Jenkinsfile]: fixed

File: Jenkinsfile (hunk 1 above)
Comment (angelozerr) [reply Jenkinsfile]: fixed

File: .github/workflows/tests.yml [hunk 11]
```diff
@@ -16,10 +16,10 @@ jobs:
         with:
           path: ~/.m2/wrapper
           key: maven-wrapper-${{ hashFiles('**/mvnw') }}
-      - name: Set up Adoptium JDK 11
+      - name: Set up Adoptium JDK 16
         uses: actions/setup-java@v2
         with:
-          java-version: '11'
+          java-version: '16'
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply .github/workflows/tests.yml]: fixed

File: .github/workflows/tests.yml [hunk 12]
```diff
@@ -16,10 +16,10 @@ jobs:
         with:
           path: ~/.m2/wrapper
           key: maven-wrapper-${{ hashFiles('**/mvnw') }}
-      - name: Set up Adoptium JDK 11
+      - name: Set up Adoptium JDK 16
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply .github/workflows/tests.yml]: fixed

File: .github/workflows/tests.yml [hunk 13]
```diff
@@ -16,10 +16,10 @@ jobs:
...
         uses: actions/setup-java@v2
         with:
-          java-version: '11'
+          java-version: '16'
           distribution: 'adopt'
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply .github/workflows/tests.yml:23]: fixed

**Comment by fbricon** [comment]
Thanks @angelozerr!

**Comment by angelozerr** [comment]
Thanks @fbricon for your great help!

**Pull request merged**